dom/events/relatedTarget.window.js

```diff
@@ -0,0 +1,85 @@
+// https://dom.spec.whatwg.org/#concept-event-dispatch
+
+const host = document.createElement("div"),
+      child = host.appendChild(document.createElement("p")),
+      shadow = host.attachShadow({ mode: "closed" }),
+      slot = shadow.appendChild(document.createElement("slot"));
+
+test(() => {
+  for (target of [shadow, slot]) {
+    for (relatedTarget of [new XMLHttpRequest(), self, host]) {
+      const event = new FocusEvent("demo", { relatedTarget: relatedTarget });
+      target.dispatchEvent(event);
+      assert_equals(event.target, null);
+      assert_equals(event.relatedTarget, null);
+    }
+  }
+}, "Reset if target pointed to a shadow tree");
+
+test(() => {
+  for (relatedTarget of [shadow, slot]) {
+    for (target of [new XMLHttpRequest(), self, host]) {
+      const event = new FocusEvent("demo", { relatedTarget: relatedTarget });
+      target.dispatchEvent(event);
+      assert_equals(event.target, null);
+      assert_equals(event.relatedTarget, null);
+    }
+  }
+}, "Reset if relatedTarget pointed to a shadow tree");
+
+async_test(t => {
```

Comment:
Can this be a sync test()?
Because this seems to be testing a synchronous event dispatch and does not need to be async_test(),
and checking the side-effect of the event handler directly e.g. on line 35.

Although it is still okay to use async_test() if you prefer as is, as it covers synchronous testing. 
I'm a bit feeling redundant that at the `t.done()` is called at the end of an async_test.

Reply:
I don't think you can use t.step stuff in a sync test.

Reply:
Yes, but as you do on line 80 (`seen = true` in the event handler to negate the value) then immediately after dispatching the event, check the value on line 83, you should be able to do the same here.
In fact, you expect the side effect of the event handler at line 35, so you don't need to wait for the `step_func()` to finish on line 39.

Reply:
I'm not waiting for it? Callbacks need to be wrapped in steps and you can only have steps in asynchronous tests, I believe. Is that not true?

Reply:
You are right, so to be clear, I'm not against `async_test` - `t.step_func` - `t.done` construct used here.

What I suggested was alternatively this test could be written as a sync test, as the callback (event handler) here is expected to be called in synchronous manner, and you can expect its side effect on the next line after `dispatchEvent`, and technically I thought it would be possible.

Probably it's a matter of taste, either isn't much better than the other.

Reply:
That does not work, since if the callback ends up throwing in that case it'd end up being catched by the framework and not the test.

Reply:
Ah, okay, I understand.  Then LGTM!